engine/src/utils/format.rs

```diff
@@ -16,23 +22,71 @@ where
         let width = self[0].as_ref().len() as u32;
         let height = self.len() as u32;
 
-        let data: Vec<u8> = self
-            .iter()
-            .flat_map(|slice| {
-                slice
-                    .as_ref()
+        let mut data = vec![];
+        {
+            let mut encoder = PngEncoder::new(&mut data);
+
+            encoder.write_image(
+                &self
                     .iter()
-                    .flat_map(|color| [color.r, color.g, color.b])
-            })
-            .collect();
+                    .flat_map(|slice| {
+                        slice
+                            .as_ref()
+                            .iter()
+                            .flat_map(|color| [color.r, color.g, color.b])
+                    })
+                    .collect::<Vec<_>>(),
+                width,
+                height,
+                image::ColorType::Rgb8,
+            );
+        }
+
+        data
+    }
+}
+
+pub trait GifFile {
+    fn to_gif(&self) -> Vec<u8>;
+}
+
+impl<Outer: ?Sized, Inner1, Inner2> GifFile for Outer
+where
+    Outer: Deref<Target = [Inner1]>,
+    Inner1: Deref<Target = [Inner2]>,
+    Inner2: AsRef<[Color]>,
+{
+    fn to_gif(&self) -> Vec<u8> {
+        let width = self[0][0].as_ref().len() as u32;
+        let height = self[0].len() as u32;
 
-        let mut image =
-            RgbImage::from_vec(width, height, data).expect("Generated image data must be valid");
+        let mut data = vec![];
+        {
+            let mut encoder = GifEncoder::new_with_speed(&mut data, 10);
```

Comment:
I know this is only for tests, but are we OK to hard code `speed` here?

Reply:
It's for preview so it doesn't matter. 10 looks nice with all animated textures in game so I kept it.

I honestly don't know how the game handles animated textures.